### PR TITLE
Refined ABI type-check on assignment

### PR DESCRIPTION
**Pull request by ahangsu**
After a few short discussions with @jasonpaulos on 2022/09/21-2022/09/22, we decided to introduce another typecheck mechanism on variable assignment for ABI values.

Such mechanism would benefit following issues and PRs: #530, #531, #537

### CODE REVIEW

File: pyteal/ast/abi/util_test.py
Comment (jasonpaulos) [inline pyteal/ast/abi/util_test.py:784]: I'd like to see a few more test cases of the existing behavior:
* `String` is not assignable to `DynamicArray[Byte]` or `DynamicBytes`, and the reverse
* `Address` is not assignable to `StaticArray[Byte, 32]` or `StaticBytes[32]`, and the reverse
* `StaticBytes[x]` is not assignable to `StaticArray[Byte, y]`, where `x != y`

File: pyteal/ast/abi/util.py
Comment (ahangsu) [reply pyteal/ast/abi/util.py:515]: For both case 1 and 2, I think we should all them, and I extended the method such that both cases can pass, in both ways of assignments (a, b or b, a).

For case 3, I add a new testcase for assigning between different `namedTuples`. Let me know if I should extend them.

File: pyteal/ast/abi/util.py
Comment (ahangsu) [reply pyteal/ast/abi/util.py:553]: This case is taken care of in latest commit.

Comment (ahangsu) [reply pyteal/ast/abi/util.py:553]: It would be captured in earlier `match`, and should not get into this position of code.

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:799]: These test cases should be added in recent commits.

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:932]: Leaving a note to document a verbal with @ahangsu and @jasonpaulos:  We discussed + agreed it'd make sense to enforce a more exhaustive exercising of `TypeSpec` assignability.  

* Rationale:  Without having to rely on a maintainer's memory alone to extend test cases, we want to defend against regressions as new `TypeSepc`s are introduced / modified.
* Caveats:  
  * We did _not_ agree upon an implementation approach.  It's understood that @ahangsu will make a pass and we can revisit.
  * There's some open questions around ease to exhaustively check _constructive_ types (e.g. `TupleTypeSpec`, `ArrayTypeSpec`).  It's understood that the solution may _not_ fully automate assignability checks.
* Related: In case helpful for brainstorming, we recently implemented a safe guard to warn us that new test cases are needed in https://github.com/algorand/pyteal/pull/519/files#diff-a1205be6b51287e45f2be7311682488d51bddf8900c13deb97df282aa090349bR396.  It's _not_ a direct fit to the problem at-hand, but it may provide inspiration.

File: pyteal/ast/abi/util.py
Comment (jasonpaulos) [reply pyteal/ast/abi/util.py:553]: Thanks for clarifying. I just want to point out that this is a potential failure point if we were to introduce new container types.

File: pyteal/ast/abi/util.py
Comment (jasonpaulos) [reply pyteal/ast/abi/util.py:515]: After more thought, I propose this rule: give types `a` and `b`, a value of type `a` is safe to view as type `b` (i.e. `type_spec_is_assignable(a,b)`) if and only if the following properties are true:
* Types `a` and `b` are encoded identically
* And type `b` is as general or more general than type `a`

So that would imply the following:

| Type `a` | Type `b` | `type_spec_is_assignable(a,b)` | reason |
| -------- | -------- | ------------------------------ | ------ |
| `DynamicArray[Byte]` | `DynamicBytes` | `true` | `DynamicBytes` is as general as `DynamicArray[Byte]` |
| `DynamicBytes` | `DynamicArray[Byte]` | `true` | `DynamicArray[Byte]` is as general as `DynamicBytes` |
| `StaticArray[Byte,N]` | `StaticBytes[N]` | `true` | `StaticBytes[N]` is as general as `StaticArray[Byte,N]` |
| `StaticBytes[N]` | `StaticArray[Byte,N]` | `true` | `StaticArray[Byte,N]` is as general as `StaticBytes[N]` |
| `String` | `DynamicBytes` | `true` | `DynamicBytes` is more general than `String` |
| `DynamicBytes` | `String` | `false` | `String` is more specific than `DynamicBytes` |
| `Address` | `StaticBytes[32]` | `true` | `StaticBytes[32]` is more general than `Address` |
| `StaticBytes[32]` | `Address` | `false` | `Address` is more specific than `StaticBytes[32]` |

Does this make sense?

Comment (ahangsu) [reply pyteal/ast/abi/util.py:515]: Good good, that makes sense to me. I just made the change to implementation and testcase.

Next up is to have a full coverage over abi object reflected by the discussion on last Friday (2022/09/30).

Comment (ahangsu) [reply pyteal/ast/abi/util.py:515]: Yes, I will take a pass and try to get it merged into my implementation.

File: pyteal/ast/abi/util.py
Comment (michaeldiamant) [reply pyteal/ast/abi/util.py:546]: Open to discussion, but want to leave the note because some time passed + we've had at least one more verbal.

Assuming there's no obvious + easy type check change, I still think it's relatively moderate cost to implement a check like https://github.com/algorand/pyteal/blob/fdf10f7c7dad4741ceec84b07413211a84086210/pyteal/ast/abi/util_test.py#L395-L404

* Since returning `true` when _not_ assignable is likely the more dangerous situation, I can imagine strengthening the analogous check by requiring the expected result to be `false` (unassignable). 
  * In the parlance of https://github.com/algorand/pyteal/pull/548/, it means looking for at least one `UnsafeAssignment` or `UnsafeBidirectional`.
* The proposal feels easier to maintainer from my perspective than a more automated check that attempts to traverse the type system hierarchy.  Obviously, the proposal leaves room for error, but I've not been able to convince myself that a better tradeoff exists.

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:1007]: @ahangsu Optional:  As an additional sanity check, rather than omitting the identity check, I see no harm in including it.  What do you think?

If you feel the change reduces clarity and/or you prefer as is, feel welcomed to resolve the thread without change.

```suggestion
            if ia == ib:
                assert abi.type_spec_is_assignable_to(a, b)
            else:
                assert not abi.type_spec_is_assignable_to(a, b)
```

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:844]: @ahangsu Request for clarification:  There's a few `SafeBidirectional` test cases that duplicate the value under test.  Examples include the highlighted line + these examples (from up above):
```
        SafeBidirectional([abi.BoolTypeSpec(), abi.BoolTypeSpec()]),
        SafeBidirectional([abi.StringTypeSpec(), abi.StringTypeSpec()]),
        SafeBidirectional([abi.AddressTypeSpec(), abi.AddressTypeSpec()]),
```

Can you help me understand how duplicating the value rather than showing a single value enriches the test?

If there's no reason to support value duplication, then I'd optionally consider changing from list to set.

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:804]: @ahangsu Testing my understanding:  Should the following test case pass?  

It fails when I tested locally.  If you think it should fail, can you help me understand the intuition?  I think the encoding is equivalent and it feels like an alias to me.

`SafeBidirectional([abi.ByteTypeSpec(), abi.Uint8TypeSpec()]),`

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:957]: @ahangsu Minor:  I think these uint cases can be added for more coverage.  Are you comfortable bringing them in?

```suggestion
    UnsafeBidirectional(
        [
            abi.Uint8TypeSpec(),
            abi.Uint16TypeSpec(),
            abi.Uint32TypeSpec(),
            abi.Uint64TypeSpec(),
        ]
    ),
    UnsafeBidirectional(
        [
            abi.ByteTypeSpec(),
            abi.Uint16TypeSpec(),
            abi.Uint32TypeSpec(),
            abi.Uint64TypeSpec(),
        ]
    ),
    UnsafeBidirectional(
```

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:949]: @ahangsu Optional:  What do you think of the following change?

* When reviewing the highlighted logic, I became a little worried that we may later add a new type and it'll _unintentionally_ go through the `return` branch.
* The proposal is to enforce that _absence_ is a deliberate decision by manually maintaining a list of types allowed to be absent.
* If you feel confident the as is approach provides sufficient coverage, then feel welcomed to resolve as is.

```suggestion
    # Abstract types and types without safe assignments should _not_ appear in test cases.
    if isabstract(ts) or ts in [
        abi.AccountTypeSpec,
        abi.ApplicationTypeSpec,
        abi.AssetTypeSpec,
        abi.BoolTypeSpec,
        abi.ByteTypeSpec,
        abi.Uint8TypeSpec,
        abi.Uint16TypeSpec,
        abi.Uint32TypeSpec,
        abi.Uint64TypeSpec,
    ]:
        assert not exists_in_safe_assignment(ts)
    else:
        assert exists_in_safe_assignment(ts)
```

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [inline pyteal/ast/abi/util_test.py:1021]: @ahangsu Optional:  In a similar vein to prior remarks, the _suggestion_ proposes ensuring test cases are deliberately omitted by inverting the assertion.   Feel welcomed to resolve if you prefer as is.

```suggestion
    if isabstract(ts):
        assert not exists_in_unsafe_bidirectional(ts)
    else:
        assert exists_in_unsafe_bidirectional(ts)
```

Comment (michaeldiamant) [reply pyteal/ast/abi/util_test.py:1021]: Additional note - If you take in the suggestion here, an analogous change can be made in `test_type_spec_is_assignable_safe_bidirectional_full_coverage`.

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:804]: This feels like something worth discussing. When we started working on ABI, even before `TypeSpec` was introduced, we did want to make `byte` inherit from `byte`, see #194.

Later that approach did not work out, and in PR #222 we introduced `TypeSpec`, but the inheritance of `byte` from `uint8` was removed, in this commit: https://github.com/algorand/pyteal/pull/222/commits/4fe7dbf1bf24c1e1b30a96291e9f0e308f70f8d0.

My thinking is, we probably should not change the implementation on `uint` side, but add a rule over `uint`, such that 2 typespecs are mutually assignable if they have same byte size.

I would also like to consult @jasonpaulos for his opinion, in case he sees some drawbacks that we failed to see.

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:1007]: I don't see the harm of including it, but so long as typespecs are identical, they should be bidirectional assignable. I would step even forward for change
```suggestion
            if ia == ib:
                assert abi.type_spec_is_assignable_to(a, b)
                assert abi.type_spec_is_assignable_to(b, a)
                continue
            assert not abi.type_spec_is_assignable_to(a, b)
```
What do you think?

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:1021]: The boundary check makes sense to me, though the abstract class cannot be instantiated, such that they naturally do not exist in test cases, but it may still worth it to add the check. I will add corresponding change in `SAFE_BIDIRECTIONAL` side.

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [inline pyteal/ast/abi/util_test.py:869]:
```suggestion
    if isabstract(ts):
        assert not exists_in_safe_bidirectional(ts)
```

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [inline pyteal/ast/abi/util_test.py:870]:
```suggestion
    if isabstract(ts):
        assert not exists_in_safe_bidirectional(ts)
    else:
        assert exists_in_safe_bidirectional(ts)
```

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:958]: sure, that makes sense to me

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:844]: moving to set is making sense, it was mistaken to have duplicate cases.

There are some difficulty in moving to set, for some type spec instances are not hashable. Need to figure out how to get it work.

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:950]: I can take this change, and I will try to explain more on comment on the choice of the set of type specs to skip.

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py:804]: I pushed the change to reflect that the `uint8` and `byte` are mutual assignable, let me know if we should roll back.

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [reply pyteal/ast/abi/util_test.py]: Makes sense to me - thanks for bringing in the change.

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [reply pyteal/ast/abi/util_test.py]: @ahangsu While it'd be nice to understand the behavior, I think https://github.com/algorand/pyteal/pull/540/commits/4315dd0d90da2c8a1e428b551ce7c0e160faf8a2 is sufficient for now.  Thanks for addressing the concern - Resolving.

File: pyteal/ast/abi/util_test.py
Comment (michaeldiamant) [reply pyteal/ast/abi/util_test.py:804]: @ahangsu Thanks for the change in https://github.com/algorand/pyteal/pull/540/commits/3256254417b66dcc8c2a3a353e196207cc8fea71.  It matches my expectation.  

@jasonpaulos Leaving the thread open for you to sanity check - Can you confirm you're on-board?

File: pyteal/ast/abi/util_test.py
Comment (jasonpaulos) [inline pyteal/ast/abi/util_test.py:924]: I'm still uncertain about allowing different `NamedTuple` subclasses to be assignable to each other. We cannot know that one is more general than the other.

Consider this example:

```python
class Point(abi.NamedTuple):
    x: abi.Field[abi.Uint64]
    y: abi.Field[abi.Uint64]

class AccountRecord(abi.NamedTuple):
    algoBalance: abi.Field[abi.Uint64]
    assetBalance: abi.Field[abi.Uint64]
```

While these two `NamedTuples` are assignable under the current implementation, they represent very different pieces of information and it would likely not be safe/intended to allow one to be viewed as the other.

The conservative option at this point would be to disallow different `NamedTuple` subclasses from being assignable to each other.

In terms of the example, I propose that `Point` and `AccountRecord` should be assignable to `abi.Tuple[abi.Uint64, abi.Uint64]`, but not to each other.

Thoughts on this?

**Review by jasonpaulos — COMMENTED** [review]
Mostly looks good, but I have concerns about `NamedTuples`

File: CHANGELOG.md
Comment (ahangsu) [reply CHANGELOG.md:4]: Sure, that makes sense. Thanks!

File: pyteal/ast/abi/util.py
Comment (ahangsu) [reply pyteal/ast/abi/util.py:504]: Pushed something for docstring in https://github.com/algorand/pyteal/pull/540/commits/312e96f74a6308062798d745823bad39e13b3549, does that look good?

Comment (jasonpaulos) [reply pyteal/ast/abi/util.py:504]: Looks great, thanks!

File: pyteal/ast/abi/util_test.py
Comment (ahangsu) [reply pyteal/ast/abi/util_test.py]: I pushed something in [d12d158](https://github.com/algorand/pyteal/pull/540/commits/d12d158dc04b27884496eb2a8d245d8a552be2e9) as the first step.

My thinking for the thread above:
> Do we want to have assignability between `NamedTuple`?

Only when the `NamedTupleTypeSpec` are *strictly identical*, or we cannot compare the generality.

> Do we still want assignability between `Tuple` and `NamedTuple`?

My preference is still keep the mutual assignability. I see scenarios that one may want to assign some `Tuple` like stuffs into more structured `NamedTuple`.

> Do we want to consider the child classes of `Point` or `AccountRecord`, i.e., child classes of child class of `NamedTuple`?

I think we want to prohibit the patterns. `NamedTuple` instances must be inherited from `NamedTuple`, and only inherited from `NamedTuple`.

Thoughts on this?

Comment (jasonpaulos) [reply pyteal/ast/abi/util_test.py]: > I think we want to prohibit the patterns. `NamedTuple` instances must be inherited from `NamedTuple`, and only inherited from `NamedTuple`.

I definitely agree on this point. I don't think we originally considered this case when we implemented `NamedTuple`, but if you were to subclass a subclass of `NamedTuple` today, you would see incorrect behavior.

Consider this example:

```python
class Point2D(abi.NamedTuple):
    """A point in 2D space."""
    x: abi.Field[abi.Uint64]
    y: abi.Field[abi.Uint64]

class Point3D(Point2D):
    """A point in 3D space."""
    z: abi.Field[abi.Uint64]
```

You would expect `Point3D` to inherit the `x` and `y` properties of `Point2D`, plus its own `z` property, but that's not actually what happens. Instead, `Point3D` would only have the `z` property.

This is because the [`get_annotations`](https://github.com/algorand/pyteal/blob/33a4e5e6691ac64f963bdcde036c2a557cf7945d/pyteal/ast/abi/tuple.py#L562) function we use to find all fields on `NamedTuple` classes _only_ looks at the fields declared in that class and ignores all inherited fields.

Due to this incorrect behavior, I'd argue that the fact you can declare and instantiate `Point3D` at all is a bug and should be prevented.

I think the preferred alternative to this example would be embedding instead of inheriting, e.g.:

```python
class Point3D(Point2D):
    """A point in 3D space."""
    point2d: abi.Field[Point2D]
    z: abi.Field[abi.Uint64]
```

Syntactically this is a little less convenient (you have to go through `point2d` to access at `x` and `y`), but it does convey the same idea.

File: pyteal/ast/abi/tuple.py
Comment (ahangsu) [reply pyteal/ast/abi/tuple.py:560]: good find, I think I prefer the second route, though `__init_subclass__` give me some hope that the misbehavior can be stopped in some form.

File: pyteal/ast/abi/tuple.py
Comment (jasonpaulos) [inline pyteal/ast/abi/tuple.py:578]: Can `issubclass` replace `__bfs_bases_tree_contain_named_tuple`?

Comment (ahangsu) [reply pyteal/ast/abi/tuple.py]: good find, take bfs down.

**Comment by jasonpaulos** [comment]
I have one final reservation. Right now `NamedTupleTypeSpec` does not override the `__eq__` method, meaning this code currently passes:

```python
class Point(abi.NamedTuple):
    x: abi.Field[abi.Uint64]
    y: abi.Field[abi.Uint64]

class AccountRecord(abi.NamedTuple):
    algoBalance: abi.Field[abi.Uint64]
    assetBalance: abi.Field[abi.Uint64]

p = Point()
ar = AccountRecord()

assert p.type_spec() == ar.type_spec()
assert abi.type_spec_is_assignable_to(p.type_spec(), ar.type_spec())
```

I'd argue having the equals method return true between different `NamedTuple` subclasses and between a `Tuple` and a `NamedTuple` is _not_ what we want.

**Comment by ahangsu** [comment]
oh boy 🤦 just pushed to save the `NamedTupleTypeSpec` equality check

File: pyteal/ast/abi/tuple.py
Comment (michaeldiamant) [inline pyteal/ast/abi/tuple.py:567]: @ahangsu Minor and optional - It seems the wording can be clarified.  I don't feel strongly.

```suggestion
        """This method ensures one only constructs directly from `NamedTuple`,
```

Comment (ahangsu) [reply pyteal/ast/abi/tuple.py:567]: 👍

**Review by michaeldiamant — APPROVED** [review]
@ahangsu Looks correct - thanks for addressing the added feedback.

Consider these approval notes:
* I optionally suggest holding for a 2nd approval.
* Once this PR is merged, I expect https://github.com/algorand/pyteal/pull/531 will be updated to leverage `type_spec_is_assignable_to` _prior to_ issuing a PyTeal release.

**Review by jasonpaulos — COMMENTED** [review]
This looks good! I only left a small comment about calling out some of the recent changes in the changelog

File: CHANGELOG.md
Comment (ahangsu) [reply CHANGELOG.md:6]: pushed the change

**Comment by ahangsu** [comment]
I just dump everything to master, and get #531 merged ASAP.

**Review by jasonpaulos — APPROVED** [review]

**Pull request merged**
